# A View of Air Travel Safety From NTSB Aviation Accident Data

### Table of Content

1. [Business Objectives](obj)
2. [Data Understanding](dataUnd)  
  2.1 [Read NTSB aviation accident data](read)  
  2.2 [Check data size]()  
  2.3 [Check field names]()  
  2.4 [Check sample data]()  
3. [Data Preparation](dataPrep)  
  3.1 [Which fields have missing data?]()  
  3.2 [Handle missing data]()  
  3.3 [Keep relevant data field]()  
  3.4 [Country counts]()  
  3.5 [Numerical fields statistics]()  
  3.6 [Handel categorical variables]()  
  3.7 [Clean data]()  
  3.8 [Verify data quality]()  
4. [Statistic Analysis](analysis)
5. [Deployment](deploy)
6. [Conclusion](conclusion)

Addionally:  
[Medium Article](https://medium.com/@qiuxinjie/3-things-you-need-to-know-about-the-of-air-travel-safety-43cddf33cd23)

### Business Objectives
<a id="obj"></a>

Safety is a concern for every air traveller. They may wonder the following questions:

1. When is the most dangerous time during each flight phases?
2. Is airplane more safe or less safe during the last few decades?
3. Which airplane has the highest fatal accidents?

National Transportation Safety Board (NTSB) provides aviation accident data.

Let's first load all the necessary libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Data Understanding
<a id="dataUnd"></a>

#### 1. Read NTSB aviation accident data

In [2]:
df = pd.read_csv('./AviationData.csv', engine='python')

#### 2. Check data size

In [3]:
num_rows = df.shape[0] #Provide the number of rows in the dataset
num_cols = df.shape[1] #Provide the number of columns in the dataset

print("there are {} rows in the dataset!".format(num_rows))
print("there are {} columns in the dataset!".format(num_cols))

there are 81013 rows in the dataset!
there are 31 columns in the dataset!


#### 3. Check field names

In [6]:
df.columns.tolist()

['Event.Id',
 'Investigation.Type',
 'Accident.Number',
 'Event.Date',
 'Location',
 'Country',
 'Latitude',
 'Longitude',
 'Airport.Code',
 'Airport.Name',
 'Injury.Severity',
 'Aircraft.Damage',
 'Aircraft.Category',
 'Registration.Number',
 'Make',
 'Model',
 'Amateur.Built',
 'Number.of.Engines',
 'Engine.Type',
 'FAR.Description',
 'Schedule',
 'Purpose.of.Flight',
 'Air.Carrier',
 'Total.Fatal.Injuries',
 'Total.Serious.Injuries',
 'Total.Minor.Injuries',
 'Total.Uninjured',
 'Weather.Condition',
 'Broad.Phase.of.Flight',
 'Report.Status',
 'Publication.Date']

#### 4. Check sample data
Look at the first few samples in the dataset to get a feeling of how the data looks like

In [7]:
df.head().T

0  \
Event.Id                           20171230X01728   
Investigation.Type                       Accident   
Accident.Number                        WPR18LA058   
Event.Date                             2017-12-29   
Location                       San Bernardino, CA   
Country                             United States   
Latitude                                  34.0917   
Longitude                                -117.248   
Airport.Code                                  SBD   
Airport.Name                  SAN BERNARDINO INTL   
Injury.Severity                         Non-Fatal   
Aircraft.Damage                       Substantial   
Aircraft.Category                        Airplane   
Registration.Number                        N222AT   
Make                                       CESSNA   
Model                                         210   
Amateur.Built                                  No   
Number.of.Engines                               1   
Engine.Type                         Reciprocating   
FAR.Description         Part 91: General Aviation   
Schedule                                      NaN   
Purpose.of.Flight                        Personal   
Air.Carrier                                   NaN   
Total.Fatal.Injuries                          NaN   
Total.Serious.Injuries                        NaN   
Total.Minor.Injuries                          NaN   
Total.Uninjured                                 1   
Weather.Condition                             VMC   
Broad.Phase.of.Flight                     LANDING   
Report.Status                         Preliminary   
Publication.Date                       2018-01-08   

                                                  1  \
Event.Id                             20171229X23639   
Investigation.Type                         Accident   
Accident.Number                          GAA18CA094   
Event.Date                               2017-12-28   
Location                              Spearfish, SD   
Country                               United States   
Latitude                                    44.4847   
Longitude                                  -103.788   
Airport.Code                                    SPF   
Airport.Name            BLACK HILLS-CLYDE ICE FIELD   
Injury.Severity                           Non-Fatal   
Aircraft.Damage                         Substantial   
Aircraft.Category                          Airplane   
Registration.Number                           N11KX   
Make                                          PIPER   
Model                                    PA 32-301T   
Amateur.Built                                    No   
Number.of.Engines                                 1   
Engine.Type                                     NaN   
FAR.Description           Part 91: General Aviation   
Schedule                                        NaN   
Purpose.of.Flight                          Personal   
Air.Carrier                                     NaN   
Total.Fatal.Injuries                            NaN   
Total.Serious.Injuries                          NaN   
Total.Minor.Injuries                            NaN   
Total.Uninjured                                   1   
Weather.Condition                               VMC   
Broad.Phase.of.Flight                           NaN   
Report.Status                           Preliminary   
Publication.Date                         2018-01-09   

                                                2                          3  \
Event.Id                           20180102X10253             20171227X12320   
Investigation.Type                       Accident                   Accident   
Accident.Number                        CEN18LA064                 GAA18CA096   
Event.Date                             2017-12-28                 2017-12-27   
Location                             Paulding, OH             Upper Lake, CA   
Country                             United States              United States   
Latitude                      

### Data Preparation
<a id="dataPrep"></a>

#### 1. Which fields have missing data?

In [8]:
df.isnull().mean()*100

Event.Id                   0.000000
Investigation.Type         0.000000
Accident.Number            0.000000
Event.Date                 0.000000
Location                   0.093812
Country                    0.625825
Latitude                  66.290595
Longitude                 66.301705
Airport.Code              43.518941
Airport.Name              40.086159
Injury.Severity            0.000000
Aircraft.Damage            3.095799
Aircraft.Category         70.097392
Registration.Number        4.115389
Make                       0.090109
Model                      0.125906
Amateur.Built              0.751731
Number.of.Engines          5.355931
Engine.Type                4.556059
FAR.Description           70.480046
Schedule                  85.646748
Purpose.of.Flight          5.216447
Air.Carrier               95.050177
Total.Fatal.Injuries      30.467950
Total.Serious.Injuries    33.390937
Total.Minor.Injuries      31.981287
Total.Uninjured           16.129510
Weather.Condition          2

About 66% data don't have Latitude/Longitude informaiton, about 40% data don't have airport information.


#### 2. Handle missing data

For Fatal.Injuries, Serious.Injuries, Minor.Injuries, Uninjured, fill missing data with 0. For Number.of.Engines, fill missing data with 1 (the majority of airplanes have only one engine)

In [9]:
df.fillna(value={'Total.Fatal.Injuries': 0,
                'Total.Serious.Injuries': 0,
                'Total.Minor.Injuries': 0,
                'Total.Uninjured': 0,
                'Number.of.Engines': 1}, 
          inplace = True)

#### 3.  Keep relevant data field
Keep only these data field relevant to the questions

In [10]:
df = df[['Accident.Number',
        'Event.Date',
        'Location',
        'Country',
        'Investigation.Type',
        'Registration.Number',
        'Make',
        'Model',
        'Amateur.Built',
        'Number.of.Engines',
        'Total.Fatal.Injuries',
        'Total.Serious.Injuries',
        'Total.Minor.Injuries',
        'Total.Uninjured',
        'Broad.Phase.of.Flight']]

#### 4. Country counts
List the number of entries for each county in the decending order

In [20]:
df['Country'].value_counts()

United States         76188
Canada                  270
Brazil                  246
United Kingdom          230
Mexico                  228
Australia               217
Bahamas                 196
France                  171
Germany                 163
Colombia                140
Spain                   114
Venezuela                93
Italy                    92
Japan                    90
Argentina                81
Indonesia                76
Peru                     74
Ecuador                  74
Russia                   68
Switzerland              67
Dominican Republic       56
Guatemala                53
Ireland                  52
New Zealand              49
China                    49
Austria                  47
Sweden                   41
Chile                    40
India                    39
Korea, Republic Of       39
                      ...  
Burma                     1
St Lucia                  1
Isle of Man               1
Anguilla                  1
Guinea              

Clearly data from the United States dominate.

#### 5. Numerical fields statistics

In [12]:
df.describe()

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       81013.000000          81013.000000            81013.000000   
mean            1.139471              0.571698                0.214583   
std             0.437485              5.198918                1.130598   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             4.000000            349.000000              111.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          81013.000000     81013.000000  
mean               0.345636         4.890919  
std                2.303591        26.925647  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

#### 6. Handel categorical variables

No need to code categorical variables since there's no maching learning modeling in this analysis

#### 7. Clean data

Notice the accident with greatest fatals have 349 fatals, let's check out what accident it is.

In [13]:
df.loc[df['Total.Fatal.Injuries']==349].T

40130             40131
Accident.Number              DCA97WA007B       DCA97WA007A
Event.Date                    1996-11-12        1996-11-12
Location                New Delhi, India  New Delhi, India
Country                            India             India
Investigation.Type              Accident          Accident
Registration.Number                  NaN               NaN
Make                             Tupolev            Boeing
Model                             TU-154           747-168
Amateur.Built                         No                No
Number.of.Engines                      1                 1
Total.Fatal.Injuries                 349               349
Total.Serious.Injuries                 0                 0
Total.Minor.Injuries                   0                 0
Total.Uninjured                        0                 0
Broad.Phase.of.Flight                NaN               NaN

It turns out it is the world's deadliest mid-air collision *1996 Charkhi Dadri mid-air collision*. The crash killed all 349 people on board both planes. Both airplanes have their entries in the NTSB dataset, but both listed 349 fatals in their entry. Saudi Arabian Airlines Boeing 747 had 312 Fatalities, Kazakhstan Airlines has 37 fatalities. Need to make correction to the falality count for both entries in the dataset.

In [14]:
df.loc[df['Accident.Number']=='DCA97WA007A','Total.Fatal.Injuries'] = 312
df.loc[df['Accident.Number']=='DCA97WA007B','Total.Fatal.Injuries'] = 37

#### 8. Verify data quality

Verify data quality by spot checking.

First look at the September 11 attacks. 4 airplanes crashed on that day.

In [16]:
df.loc[df['Event.Date']=='2001-09-11'].T

29526          29527              29528  \
Accident.Number              DCA01MA065     DCA01MA064         DCA01MA063   
Event.Date                   2001-09-11     2001-09-11         2001-09-11   
Location                Shanksville, PA  Arlington, VA  New York City, NY   
Country                   United States  United States      United States   
Investigation.Type             Accident       Accident           Accident   
Registration.Number              N591UA         N644AA             N612UA   
Make                             Boeing         Boeing             Boeing   
Model                               757        757-200          767-200ER   
Amateur.Built                        No             No                 No   
Number.of.Engines                     1              1                  2   
Total.Fatal.Injuries                 44             64                 65   
Total.Serious.Injuries                0              0                  0   
Total.Minor.Injuries                  0              0                  0   
Total.Uninjured                       0              0                  0   
Broad.Phase.of.Flight            CRUISE         CRUISE             CRUISE   

                                    29529  
Accident.Number                DCA01MA060  
Event.Date                     2001-09-11  
Location                New York City, NY  
Country                     United States  
Investigation.Type               Accident  
Registration.Number                N334AA  
Make                               Boeing  
Model                           767-200ER  
Amateur.Built                          No  
Number.of.Engines                       2  
Total.Fatal.Injuries                   92  
Total.Serious.Injuries                  0  
Total.Minor.Injuries                    0  
Total.Uninjured                         0  
Broad.Phase.of.Flight              CRUISE

Then list accidents with more than 250 fatalities:

In [17]:
df.loc[df['Total.Fatal.Injuries']>250].T

5420              29244             40131  \
Accident.Number               DCA14RA127        DCA02MA001       DCA97WA007A   
Event.Date                    2014-07-17        2001-11-12        1996-11-12   
Location                Hrabove, Ukraine  Belle Harbor, NY  New Delhi, India   
Country                          Ukraine     United States             India   
Investigation.Type              Accident          Accident          Accident   
Registration.Number                  NaN            N14053               NaN   
Make                              BOEING  Airbus Industrie            Boeing   
Model                          777 - 206       A300B4-605R           747-168   
Amateur.Built                         No                No                No   
Number.of.Engines                      1                 2                 1   
Total.Fatal.Injuries                 295               265               312   
Total.Serious.Injuries                 0                 0                 0   
Total.Minor.Injuries                   0                 0                 0   
Total.Uninjured                        0                 0                 0   
Broad.Phase.of.Flight                NaN           TAKEOFF               NaN   

                                            58931           67415  
Accident.Number                        DCA89RA014      DCA86RA010  
Event.Date                             1988-12-21      1985-12-12  
Location                LOCKERBIE, United Kingdom  GANDER, Canada  
Country                            United Kingdom          Canada  
Investigation.Type                       Accident        Accident  
Registration.Number                        N739PA          N950JW  
Make                                       BOEING         DOUGLAS  
Model                                     747-121         DC-8-63  
Amateur.Built                                  No              No  
Number.of.Engines                               4               4  
Total.Fatal.Injuries                          270             256  
Total.Serious.Injuries                          2               0  
Total.Minor.Injuries                            3               0  
Total.Uninjured                                 0               0  
Broad.Phase.of.Flight                         NaN             NaN

### Statistic Analysis
<a id="analysis"></a>

Look for the counts of each unique values (including NAN) in Investigation Type

In [ ]:
vals = df['Investigation.Type'].value_counts(normalize=True, dropna=False)[:10]*100
vals.plot(kind="bar");

Majority of them are accident, only a small a fraction is incident. 
Look at country again, list the top 10 countries in the dataset:

In [ ]:
vals = df['Country'].value_counts(normalize=True, dropna=False)[:10]*100
vals.plot(kind="bar");
plt.grid()

More than 90% data are from the United States, followed by Canada, Brazil, United Kindown, and Mexico.

Now let's focus on these accident in the United States with at least one fatality, from data after Jan. 1982. (Data from before this month is not complete).

In [ ]:
df['Date'] = pd.to_datetime(df['Event.Date'])
df.drop('Event.Date', axis=1, inplace=True)
df_fatal = df.loc[(df['Total.Fatal.Injuries']>1) &
           (df['Date']>='1982-01') &
           (df['Country'] == 'United States'),
            :]

How each aircaft model contributes to the fatal accidents.

In [ ]:
vals = df_fatal['Make'].value_counts(normalize=True, dropna=False)[:20]*100
vals.plot(kind="bar");
plt.grid()

There is issue here, "CESSNA" (all CAPITALIZED) and "Cessna" (Only first letter Capitalized) has been listed as two separate models, so is "PIPER" and "Piper". Let's combined them by change all of them to upper case.

In [ ]:
vals = df_fatal['Make'].str.upper().value_counts(normalize=True, dropna=False)[:20]*100
plt.figure(figsize=(8,6))
vals.plot(kind="bar");
plt.grid()
plt.title("Accident by aircraft models")
plt.show()

About 29% fatal accidents from Cessna, another 22% from Piper. 12% from Beech. It is not a surprise since Cessna, Piper Aircraft, and Beechcraft were considered the "Big Three" in the field of general aviation manufacturing.

In [ ]:
vals = df_fatal['Amateur.Built'].value_counts(normalize=True, dropna=False)*100
plt.figure(figsize=(8,6))
vals.plot(kind="bar");
plt.grid()
plt.title("Accident by Amateur Built")
plt.show()

In [ ]:
vals = df_fatal['Number.of.Engines'].value_counts(normalize=True, dropna=False)*100
plt.figure(figsize=(8,6))
vals.plot(kind="bar");
plt.grid()
plt.title("Accident by number of engines")
plt.show()

Check the fatal accident and fatality rate during each flight phase

In [ ]:
vals = df_fatal['Broad.Phase.of.Flight'].value_counts(normalize=True, dropna=False)*100
plt.figure(figsize=(8,6))
vals.plot(kind="bar");
plt.grid()
plt.title("Fatal accidents during each flight phase")
plt.xlabel("Flight phase")
plt.ylabel("Fatal accident percentage during each flight phase")
plt.show()

In [ ]:
vals = df_fatal.groupby(by='Broad.Phase.of.Flight')['Total.Fatal.Injuries'].sum().sort_values(ascending=False)
plt.figure(figsize=(8,6))
(vals/vals.sum()*100).plot(kind="bar");
plt.grid()
plt.title("Fatality during each flight phase")
plt.xlabel("Flight phase")
plt.ylabel("Fatality percentage during each flight phase")
plt.show()

The year to year Fatality rate

In [ ]:
fatal_agg = df_fatal['Total.Fatal.Injuries'].groupby([df_fatal['Date'].dt.year]).agg({'sum','count'})
fatal_agg

plt.figure(figsize=(8,6))
plt.plot(fatal_agg[['sum','count']], 'o-', lw=4, ms=10 )
plt.legend(('Fatal counts', 'Fatal accidents'))
plt.title('Fatal accident historical trend')
plt.xlabel('Year')
plt.ylabel('Count')
plt.grid()
plt.show()

In [ ]:
vals = df_fatal.groupby(by='Number.of.Engines')['Total.Fatal.Injuries'].sum().sort_values(ascending=False)
plt.figure(figsize=(8,6))
(vals/vals.sum()*100).plot(kind="bar");
plt.grid()
plt.title("Fatality for each engine numbers")
plt.xlabel("Number of engines")
plt.ylabel("Fatality percentage for each engine numbers")
plt.show()

## Deployment
<a id="deploy"></a>

The audience can find the the result of this analysis in the following locations:  
- [Medium article](https://medium.com/@qiuxinjie/3-things-you-need-to-know-about-the-of-air-travel-safety-43cddf33cd23)
- [README on GitHub](https://github.com/tjphoton/Aviation-Safety)
- [Jupyter Notebook on GitHub](https://github.com/tjphoton/Aviation-Safety/blob/master/Aviation.ipynb)

## Conclusion
<a id="conclusion"></a>

In this analysis, we took a look at the aviation safety is according to NTSB data.


1. Maneuvering, cruise, takeoff, and approach are the four most dangerous flight phases causing fatality.
2. Aviation fatal incident events are less and less over the years, aviation fatality rate is also declining, indicate the trend of air travel safety.
3. Cessna, Piper, Beech are the there most popular general aviation aircraft manufactures, making them the most fatality accident related aircraft models. But you should be less concerned about these general aviation aircraft models when you travel with airlines.
